src/main/java/com/ace/alfox/game/MoveAction.java

```diff
@@ -34,6 +39,14 @@ public ActionResult applyAction(Player player, Map<String, Object> params) {
                                     // maybe do running away mechanic here?
         player.location = player.location.add(directions.get(direction));
         result.log("traveled " + direction);
+
+        Location location = db.locations.find(eq("coordinates", player.location)).firstOrDefault();

```

Comment:
Would love to see an intermediate layer. `DataRepository` handles all requests for data (whether it's from a cache, database, network, file system, etc). It would have methods like `dataRepository.findLocation(player.location)` and the calling code, `MoveAction` doesn't care about whether it's from Database or anywhere else. Not a blocker for this review, but wanted to get your thoughts on that! :)

I think this would make it really easy to swap out the database (say from LevelDB to Nitrite) without having to change a lot of code, just the DataRepository.

Reply:
Ha, I was totally thinking this while coding. The Nitrite API is nice, but I totally need to wrap all the repositories with custom fetching logic. The worst bit is that the Nitrate `ObjectRepository` is an interface, so I can't just extend it, I'd need to implement all of the minutia. So it's totally something that needs to happen, I'm just not looking forward to it :P

Reply:
For large interfaces I always provide an adapter. ie a class that implements all the methods and does noops, so if you just want to override one method you can do that. Maybe we could add one for ObjectRepository to make it easier to use.

Taking a step back, I don't think you'd have to do anything different to have the DataRepository layer. You would have this code call `dataRepository.findLocation(player.location)` and then datarepo would look like this:

```java
class DataRepository {
    // init stuff

    Location findLocation(Location location) {
       return db.locations.find(eq("coordinates", player.location)).firstOrDefault();
    }
```

I think this is fine. We would need to change all the `eq()` and stuff if we swapped out Nitrite, but it's all in one place and could be replaced slowly. ie you could add a second DB and then replace endpoints as necessary.